cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/QueryPlanTest.scala

```diff
@@ -441,16 +836,287 @@ class QueryPlanTest extends DocumentingTestBase with SoftReset {
+          |that check for the existence of a pattern.
+          |First, the normal expression predicate is evaluated, and, only if it returns `false`, is the costly pattern predicate evaluated.""".stripMargin,
+      queryText =
+        """MATCH (other:Person)
+          |WHERE other.age > 25 OR (other)-[:FRIENDS_WITH]->(:Person)
+          |RETURN other.name""".stripMargin,
+      assertions = (p) => assertThat(p.executionPlanDescription().toString, containsString("SelectOrSemiApply"))
+    )
+  }
+
+  @Test def selectOrAntiSemiApply() {
+    profileQuery(
+      title = "Select Or Anti Semi Apply",
+      text =
+        """The `SelectOrAntiSemiApply` operator tests for the absence of a pattern predicate and evaluates a predicate.""".stripMargin,
+      queryText =
+        """MATCH (other:Person)
+          |WHERE other.age > 25 OR NOT (other)-[:FRIENDS_WITH]->(:Person)
+          |RETURN other.name""".stripMargin,
+      assertions = (p) => assertThat(p.executionPlanDescription().toString, containsString("SelectOrAntiSemiApply"))
+    )
+  }
+
+  @Test def conditionalApply() {
+    profileQuery(
+      title = "Conditional Apply",
+      text =
+        """The `ConditionalApply` operator checks whether a variable is not `null`, and if so, the right child operator will be executed.""".stripMargin,
+      queryText =
+        """MERGE (p:Person {name: 'Andres'})
+          |ON MATCH SET p.exists = true""".stripMargin,
+      assertions = (p) => assertThat(p.executionPlanDescription().toString, containsString("ConditionalApply"))
+    )
+  }
+
+  @Test def antiConditionalApply() {
+    profileQuery(
+      title = "Anti Conditional Apply",
+      text =
+        """The `AntiConditionalApply` operator checks whether a variable is `null`, and if so, the right child operator will be executed.""".stripMargin,
+      queryText =
+        """MERGE (p:Person {name: 'Andres'})
+          |ON CREATE SET p.exists = true""".stripMargin,
+      assertions = (p) => assertThat(p.executionPlanDescription().toString, containsString("AntiConditionalApply"))
+    )
+  }
+
+  @Test def assertSameNode() {
+    profileQuery(
+      title = "Assert Same Node",
+      text =
+        """The `AssertSameNode` operator is used to ensure that no uniqueness constraints are violated.
+          |The example looks for the existence of a team with the supplied name and id, and if one does not exist,
+          |it will be created. Owing to the existence of two uniqueness constraints
+          |on `:Team(name)` and `:Team(id)`, any node that would be found by the `UniqueIndexSeek`
+          |must be the very same node, or the constraints would be violated.
+        """.stripMargin,
+      queryText =
+        """MERGE (t:Team {name: 'Engineering', id: 42})""".stripMargin,
+      assertions = (p) => assertThat(p.executionPlanDescription().toString, containsString("AssertSameNode"))
+    )
+  }
+
+  @Test def nodeHashJoin() {
```

Comment:
Hash Joins General
-=-=-=-=-=-=-=-=-=-
Hash joins have two inputs: the build input and probe input. The query planner assigns these roles so that the smaller of the two inputs is the build input. The build input is pulled in eagerly and used to build a probe table. Once this is finished, the probe table is checked for each row coming from the probe input side.

In query plans, the build input is always the left operator, and the probe input the right operator.

Hash Join Operators:

Node Hash Join:
A variation that does the hash join on node ids, which means it can be done in a faster way, using primitive types and arrays.

Value Hash Join:
Allows for arbitrary values to be used as the join key. Most often used to solve predicates of the form equality predicates between two property columns.

Outer Hash Join:
Instead of discarding rows that are not found in the probe table, will yield a single row with null instead.

Reply:
👍